File: project_euler/problem_723/solution.py
"""
Problem 723: Pythagorean Quadrilaterals

Count pythagorean lattice grid quadrilaterals inscribed in circles.
S(n) = sum_{d|n} f(sqrt(d)) where f(r) counts distinct pythagorean
lattice quadrilaterals with circumradius r.
"""

def r2(n):
    """Count representations of n as sum of two squares (order matters, signs matter)."""
    count = 0
    for x in range(int(n**0.5) + 1):
        y2 = n - x*x
        if y2 < 0:
            break
        y = int(y2**0.5)
        if y*y == y2:
            if x == 0 or y == 0:
                count += 2 if (x == 0) != (y == 0) else 1
                count += 2 if (x == 0) != (y == 0) else 1
            else:
                count += 4
            if x == y and y*y == y2:
                pass  # already counted
    # Simpler: direct enumeration
    count = 0
    for x in range(-int(n**0.5)-1, int(n**0.5)+2):
        for y in range(-int(n**0.5)-1, int(n**0.5)+2):
            if x*x + y*y == n:
                count += 1
    return count

def lattice_points(d):
    """Return list of (x,y) with x^2 + y^2 = d."""
    pts = []
    s = int(d**0.5) + 1
    for x in range(-s, s+1):
        y2 = d - x*x
        if y2 < 0:
            continue
        y = int(y2**0.5 + 0.5)
        if y*y == y2:
            if y > 0:
                pts.append((x, y))
                pts.append((x, -y))
            elif y == 0:
                pts.append((x, 0))
    return pts

def f_brute(d):
    """Brute force count of pythagorean lattice grid quadrilaterals for radius sqrt(d)."""
    pts = lattice_points(d)
    n = len(pts)
    count = 0
    # Check all ordered 4-tuples (with cyclic/reflection equivalence)
    for i in range(n):
        for j in range(n):
            if j == i: continue
            for k in range(n):
                if k in (i, j): continue
                for l in range(n):
                    if l in (i, j, k): continue
                    p1, p2, p3, p4 = pts[i], pts[j], pts[k], pts[l]
                    a2 = (p1[0]-p2[0])**2 + (p1[1]-p2[1])**2
                    b2 = (p2[0]-p3[0])**2 + (p2[1]-p3[1])**2
                    c2 = (p3[0]-p4[0])**2 + (p3[1]-p4[1])**2
                    d2 = (p4[0]-p1[0])**2 + (p4[1]-p1[1])**2
                    if a2 + b2 + c2 + d2 == 8*d:
                        count += 1
    # Each quadrilateral counted 8 times (4 rotations x 2 reflections)
    # But degenerate cases need care
    return count // 8

# Verify small cases
print(f"f(1) = {f_brute(1)}")   # Expected: 1
print(f"f(2) = {f_brute(2)}")   # Expected: 1
print(f"f(5) = {f_brute(5)}")   # Expected: 38

File: project_euler/problem_723/solution.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * Problem 723: Pythagorean Quadrilaterals
 *
 * Count pythagorean lattice grid quadrilaterals inscribed in circles.
 * A quadrilateral ABCD with vertices on x^2+y^2=d is pythagorean if
 * a^2 + b^2 + c^2 + d^2 = 8d (sides squared sum to 8r^2).
 *
 * S(n) = sum_{d|n} f(sqrt(d))
 *
 * The approach uses:
 * 1. Enumerate lattice points on circle via sum-of-two-squares
 * 2. Count valid quadrilaterals satisfying the pythagorean condition
 * 3. Multiplicative structure for the large input
 */

vector<pair<int,int>> lattice_points(int d) {
    vector<pair<int,int>> pts;
    int s = (int)sqrt((double)d) + 1;
    for (int x = -s; x <= s; x++) {
        int y2 = d - x*x;
        if (y2 < 0) continue;
        int y = (int)round(sqrt((double)y2));
        if (y*y == y2) {
            pts.push_back({x, y});
            if (y > 0) pts.push_back({x, -y});
        }
    }
    return pts;
}

int main() {
    // Verify small cases
    for (int d : {1, 2, 5}) {
        auto pts = lattice_points(d);
        int n = pts.size();
        int count = 0;
        for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++) if (j != i)
        for (int k = 0; k < n; k++) if (k != i && k != j)
        for (int l = 0; l < n; l++) if (l != i && l != j && l != k) {
            auto [x1,y1] = pts[i]; auto [x2,y2] = pts[j];
            auto [x3,y3] = pts[k]; auto [x4,y4] = pts[l];
            long long a2 = (long long)(x1-x2)*(x1-x2) + (long long)(y1-y2)*(y1-y2);
            long long b2 = (long long)(x2-x3)*(x2-x3) + (long long)(y2-y3)*(y2-y3);
            long long c2 = (long long)(x3-x4)*(x3-x4) + (long long)(y3-y4)*(y3-y4);
            long long d2 = (long long)(x4-x1)*(x4-x1) + (long long)(y4-y1)*(y4-y1);
            if (a2 + b2 + c2 + d2 == 8LL * d) count++;
        }
        printf("f(sqrt(%d)) = %d\n", d, count / 8);
    }
    return 0;
}
